File: Data Structures & Algorithms/palindrome-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };

 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* curr = head;
        vector<int> l1;
        
        while(curr) {
            l1.push_back(curr->val);
            curr = curr->next;
        }
        int p1 = 0;
        int p2 = l1.size()-1;
        
        while(p1 < p2) {
            if(l1[p1] != l1[p2]) {
                return false;
            }
            p1++;
            p2--;
        }
        return true;
    }
};

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        # leaf nodes get set to nullptr
        if not root:
            return root
        
        if root.val > key:
            root.left = self.deleteNode(root.left, key)
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        else:
            if not root.right:
                return root.left
            elif not root.left:
                return root.right
            else:
                # find our in order predecessor (left subtrees greatest node)
                curr = root.left
                while curr.right:
                    curr = curr.right
                root.val = curr.val
                root.left = self.deleteNode(root.left, curr.val)
        return root

        # nodes that have one child get set to that child
        # nodes that have two children get set to one of the children
        # nodes with grandchildren cannot arbitrarily choose a child to be replaced by
        # we must keep the BST balanced
        # we either have to choose the next in-order predecessor or the in order successor
